[llvm][CodeGen] Resolve issues when updating live intervals in window scheduler

huaatian opened the pull request:
Corrupted live interval information can cause window scheduling to crash in some cases. 
By adding the missing MBB's live interval information in the ModuloScheduleExpander, 
the information can be correctly analyzed in the window scheduler.

huaatian committed:
[llvm][CodeGen] Reanalyze corrupted live intervals in window scheduling

Corrupted live interval information can cause window scheduling to crash
in some cases. Therefore, it is necessary to reanalyze this information
in such situations.

huaatian commented:
This bug was reported in https://github.com/llvm/llvm-project/pull/99454.
The root cause of the issue is that slot index and live interval information are required when constructing TripleMBB. During the analysis of some virtual registers, references are made to an MBB that has already been processed by the machine pipeliner. Since the machine pipeliner does not preserve live interval and slot index information, the following error occurs:
![image](https://github.com/user-attachments/assets/5f240511-235d-44c3-8989-d21de25a9051)
The probability of this situation occurring is low. Typically, when updating the LiveIntervals in TripleMBB, scheduled MBBs are not referenced. Therefore, in such cases, we reanalyze the live intervals.
PS :This error scenario is difficult to reproduce, so we can only try to simplify the reported scenario as much as possible.

huaatian commented:
Hi, could you please take some time to review this PR? Sorry for taking up your time, and thank you! @arsenm

huaatian commented:
ping

huaatian commented:
hi，@arsenm，could you please take a look at this PR？Thanks for your time!

arsenm commented on llvm/lib/CodeGen/WindowScheduler.cpp:
```diff
@@ -644,6 +645,20 @@ void WindowScheduler::expand() {
+      if (!SlotIndex->hasIndex(MI)) {
+        // The slot index and live intervals of MF have been corrupted and need
+        // to be reanalyzed.
+        SlotIndex->reanalyze(*MF);
+        Context->LIS->reanalyze(*MF);
```
This is much too heavy of a hammer. You should be updating specific intervals as you change them (probably need to add some handleMove calls somewhere)

huaatian replied:
You are right. 
The issue occurs in the ModuloScheduleExpander module. We have avoided the current problem by adding the minimal amount of live interval information, without affecting the window scheduler.

arsenm commented on llvm/test/CodeGen/Hexagon/swp-ws-live-intervals.mir:
```diff
@@ -0,0 +1,204 @@
+# The bug was reported at https://github.com/llvm/llvm-project/pull/99454. 
+# It is caused by the corruption of live intervals in certain scenarios.
+#
+# CHECK: Best window offset is {{[0-9]+}} and Best II is {{[0-9]+}}.
+# CHECK: Best window offset is {{[0-9]+}} and Best II is {{[0-9]+}}.
```
Better check would check the actual scheduling result

huaatian replied:
Updated

huaatian committed:
[llvm][CodeGen] Resolve issues when updating live intervals in window scheduler

Corrupted live interval information can cause window scheduling to crash
in some cases. By adding the missing MBB's live interval information in
the ModuloScheduleExpander, the information can be correctly analyzed
in the window scheduler.

arsenm reviewed (approved):
I'm surprised this is enough to update everything

huaatian commented:
> I'm surprised this is enough to update everything

This patch only adds the MBB information to the maps. 
In the window scheduler, we have already call the repairIntervalsInRange() interface to fix the missing live interval information. 
The issue occurred because the live interval information corresponding to the newly generated code after pipeliner scheduling was corrupted. 
By simply adding the MBB information, the repairIntervalsInRange() interface will no longer fail.

huaatian commented:
/cherry-pick

huaatian commented:
/cherrypick [43ba109]

huaatian commented:
/cherrypick https://github.com/llvm/llvm-project/commit/43ba1097ee747b4ec5e757762ed0c9df6255a292

huaatian commented:
/cherry-pick https://github.com/llvm/llvm-project/commit/43ba1097ee747b4ec5e757762ed0c9df6255a292
